MBBSEmu/Session/LineBreaker.cs

```diff
@@ -0,0 +1,197 @@
+
+        /// <summary>
+        /// A mapping between lineBuffer to rawBuffer.
+        ///
+        /// Each index in this array (which is a printable character) maps to
+        /// the data area where the character is in rawBuffer.
+        /// </summary>
+        private readonly int[] lineBufferToRawBuffer = new int[MAX_LINE];
+
+        /// <summary>
+        /// Contains all the bytes accumulated by parsing outgoing data,
+        /// including ANSI characters. This array is output to our clients.
+        /// </summary>
+        private readonly byte[] rawBuffer = new byte[MAX_OUTPUT_BUFFER];
+        /// <summary>
+        /// How many bytes are valid in rawBuffer
+        /// </summary>
+        private int rawBufferLength = 0;
+
+        /// <summary>
+        /// Breaks buffer into lines and calls SendToClientMethod afterwards.
+        /// </summary>
+        /// <param name="buffer">Raw output buffer going to a client</param>
+        public void SendBreakingIntoLines(byte[] buffer)
+        {
+            foreach(byte b in buffer)
+            {
+                rawBuffer[rawBufferLength++] = b;
+
+                switch (_parseState)
+                {
+                     case ParseState.APS_ESCAPE when b == '[':
+                        _parseState = ParseState.APS_ESCAPE_BRACKET;
+                        break;
+                     case ParseState.APS_ESCAPE when b != ESCAPE:
+                        _parseState = ParseState.APS_NORMAL;  // this happens if we receive an escape not followed by [, which shouldn't happen
+                        break;
+                    case ParseState.APS_ESCAPE_BRACKET when Char.IsLetter((char) b): // ANSI tags are terminated with letters
+                        _parseState = ParseState.APS_NORMAL;
+                        break;
+                    case ParseState.APS_NORMAL:
+                        switch (b)
+                        {
+                            case (byte) '\r':
+                                lineBufferLength = 0; // line ended, erase accumulated data in lineBuffer
+                                break;
+                            case ESCAPE: // escape
+                                _parseState = ParseState.APS_ESCAPE;
+                                break;
+                            case (byte) '\n':  // ignore
+                                break;
+                            default:
+                                lineBuffer[lineBufferLength] = b;
+                                lineBufferToRawBuffer[lineBufferLength++] = rawBufferLength - 1;
+                                // overflow to the next line
+                                if (lineBufferLength > TerminalColumns)
+                                {
+                                    doLineBreak(b);
+                                }
+                                break;
+                        }
+                        break;
+                }
```

Comment:
Probably add a `default` case to throw an exception if it falls through for whatever reason

Reply:
default is intended, so just commented